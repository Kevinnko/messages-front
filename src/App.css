.container {
  max-width: 1024px;
  margin: auto;
}
.message {
  display: flex;
  justify-content: center;
  align-content: center;
}
