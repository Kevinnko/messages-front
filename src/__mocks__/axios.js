// "use strict";
module.exports = {
  get: () => {
    return Promise.resolve({
      data: [
        {
          message: "Salut, comment Ã§a va ?",
          public: "false"
        },
        {
          message: "TrÃ¨s bien merci, et toi ?",
          public: "false"
        }
      ]
    });
  },
  post: (url, data) => {
    return Promise.resolve({
      data: {
        message: data.message,
        public: data.public
      }
    });
  }
};
