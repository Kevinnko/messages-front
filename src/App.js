import React from "react";
import axios from "axios";
import "./App.css";

class App extends React.Component {
  state = {
    messages: [],
    message: "",
    status: "true"
  };

  async componentDidMount() {
    var element = document.getElementsByClassName("scrollable-div")[0];
    console.log("in componentDidMount", element);

    // loading messages from api
    const response = await axios.get("http://localhost:3001/get-messages");
    this.setState({
      messages: response.data
    });
  }
  componentDidUpdate() {
    var element = document.getElementsByClassName("scrollable-div")[0];
    console.log("in componentDidUpdate", element);
    element.scrollIntoView(false);
  }

  handleChange = event => {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  };

  handleSubmit = async event => {
    event.preventDefault();

    const response = await axios.post("http://localhost:3001/post-message", {
      message: this.state.message,
      public: this.state.status
    });

    this.setState({
      messages: response.data,
      message: ""
    });
  };

  render() {
    const { messages } = this.state;

    return (
      <div className="container">
        {/* S'il y a des messages, on les affiche, sinon on affiche "pas de message" */}
        {messages.length === 0 ? (
          <div>Pas de messages</div>
        ) : (
          <MessagesList messages={messages} />
        )}
        <form
          className="form"
          style={{ display: "flex" }}
          onSubmit={this.handleSubmit}
        >
          <textarea
            className="message-input"
            placeholder="Ecrivez votre message..."
            name="message"
            cols="30"
            rows="1"
            value={this.state.message}
            onChange={this.handleChange}
          />
          <select
            className="select"
            value={this.state.status}
            name="status"
            onChange={this.handleChange}
          >
            <option value="true">Public</option>
            <option value="false">Privé</option>
          </select>
          <input
            disabled={this.state.message === "" ? true : false}
            className="button"
            type="submit"
            value="Envoyer"
          />
        </form>
      </div>
    );
  }
}
// created stateless component
export const MessagesList = props => {
  return (
    <div className="scrollable-div">
      {props.messages.map((message, index) => {
        return (
          <>
            <div key={index} className="message">
              <div className="msg-text">{message.message}</div>
              {message.public === "true" ? (
                <div className="msg-status">Public</div>
              ) : (
                <div className="msg-status">Privé</div>
              )}
            </div>
          </>
        );
      })}
    </div>
  );
};

export default App;
