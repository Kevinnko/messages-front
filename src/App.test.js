import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import axios from "axios";
import renderer, { create } from "react-test-renderer";
import { configure, mount, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

jest.mock("axios");

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe("<App /> not connected to server", () => {
  it("matches the snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should contain a paragraph telling to launch server", () => {
    const wrapper = shallow(<App />);
    const text = wrapper.find("p").text();
    expect(text).toEqual("Pas de messages : lancer le serveur.");
  });

  it("should contain one textarea, one select and one input button", () => {
    const wrapper = shallow(<App />);
    const inputs = wrapper.find("input");
    const textareas = wrapper.find("textarea");
    const selects = wrapper.find("select");
    expect(inputs.length).toEqual(1);
    expect(textareas.length).toEqual(1);
    expect(selects.length).toEqual(1);
  });
});

describe("<App /> connected to server", () => {
  // it("should call an API", async () => {
  //   const response = {
  //     data: [
  //       { message: "Coucou", status: "true" },
  //       { message: "Coucou", status: "false" }
  //     ]
  //   };
  //   expect(Array.isArray(response.data)).toEqual(true);
  //   axios.get.mockResolvedValue(response);
  //   const component = create(<App />);
  //   const instance = component.getInstance();
  //   await instance.componentDidMount();
  //   // await instance.componentDidUpdate();
  // });
});

// describe("<App />", () => {
//   it("should contain an array of objects with specific keys", async () => {
//     const response = {
//       data: [
//         { message: "Coucou", status: "true" },
//         { message: "Coucou", status: "false" }
//       ]
//     };
//     expect(Array.isArray(response.data)).toEqual(true);
//     axios.get.mockResolvedValue(response);
//     const component = create(<App />);
//     const instance = component.getInstance();
//     await instance.componentDidMount();

//     instance.state.messages.forEach(obj => {
//       expect(typeof obj).toEqual("object");

//       // chaque objet du tableau de messages doit contenir les clés 'message' et 'status
//       expect(Object.keys(obj).sort()).toEqual(["message", "status"]);

//       // Valider le type des valeurs de chaque clé
//       expect(typeof obj.message).toEqual("string");
//       expect(typeof obj.status).toEqual("string");
//     });
//   });

//   // tester le conditional public / privé
//   // tester le onSubmit
// });

// // le composant MessageList doit contenir une liste de messages
// describe("MessageList", () => {
//   it("should contain divs", () => {
//     const messages = [
//       { message: "Coucou", status: "true" },
//       { message: "Coucou", status: "false" }
//     ];

//     const component = mount(<MessagesList messages={messages} />);
//     const value = component.find("div.message");
//   });
// });
