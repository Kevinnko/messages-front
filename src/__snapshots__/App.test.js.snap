// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches the snapshot 1`] = `
<div
  className="container"
>
  <div
    style={
      Object {
        "padding": 10,
      }
    }
  >
    <p>
      Pas de messages : lancer le serveur.
    </p>
  </div>
  <form
    className="form"
    onSubmit={[Function]}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <textarea
      className="message-input"
      cols="30"
      name="message"
      onChange={[Function]}
      placeholder="Ecrivez votre message..."
      rows="1"
      value=""
    />
    <select
      className="select"
      name="status"
      onChange={[Function]}
      value="true"
    >
      <option
        value="true"
      >
        Public
      </option>
      <option
        value="false"
      >
        Privé
      </option>
    </select>
    <input
      className="button"
      disabled={true}
      type="submit"
      value="Envoyer"
    />
  </form>
</div>
`;

exports[`<App /> not connected to server matches the snapshot 1`] = `
<div
  className="container"
>
  <div
    style={
      Object {
        "padding": 10,
      }
    }
  >
    <p>
      Pas de messages : lancer le serveur.
    </p>
  </div>
  <form
    className="form"
    onSubmit={[Function]}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <textarea
      className="message-input"
      cols="30"
      name="message"
      onChange={[Function]}
      placeholder="Ecrivez votre message..."
      rows="1"
      value=""
    />
    <select
      className="select"
      name="status"
      onChange={[Function]}
      value="true"
    >
      <option
        value="true"
      >
        Public
      </option>
      <option
        value="false"
      >
        Privé
      </option>
    </select>
    <input
      className="button"
      disabled={true}
      type="submit"
      value="Envoyer"
    />
  </form>
</div>
`;
